src/cursor.rs

```diff
@@ -0,0 +1,116 @@
+use std::ops::{Add, AddAssign, Index, IndexMut, Sub, SubAssign};
+
+// TODO: AddAssign
+pub struct CursorMut<'a, T> {
+    data: &'a mut [T],
+    index: usize,
+}
+
+impl<'a, T> CursorMut<'a, T> {
+    pub fn new(data: &'a mut [T]) -> Self {
+        CursorMut { data, index: 0 }
+    }
+
+    pub fn as_slice(&self) -> &[T] {
+        &self.data[self.index..]
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        &mut self.data[self.index..]
+    }
+
+    pub fn as_ptr(&self) -> *const T {
+        self.as_slice().as_ptr()
+    }
+
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        self.as_mut_slice().as_mut_ptr()
+    }
+
+    pub fn clone<'b>(&'b mut self) -> CursorMut<'b, T> {
+        CursorMut {
+            data: self.data,
+            index: self.index,
+        }
+    }
+}
+
+impl<'a, T> Add<usize> for CursorMut<'a, T> {
```

Comment:
`impl Add` and `impl Sub` this way doesn't seem that intuitive/idiomatic in Rust.  Do you think we should do it this way?  It is more similar to the C way, but I worry that it's easy to overlook hidden operations in doing this.  When optimizing code to avoid bounds checks later, for example, it'd be more difficult to see where they occur.  What might be more idiomatic is to allow slicing with a range like normal slices do, but I haven't fully thought that out yet.

Reply:
Do we want to keep the Rust code as similar looking to the C code?  Operator overloading isn't normally done like this in Rust.  It's also harder to change when it's done this for, vs. using a method, like `.offset`, that's much easier to refactor.

Reply:
In general we want to keep the Rust code similar to the C as much as possible. In some cases we can convert the pointer arithmetic operations into safe slice operations, but I don't think that's always going to be possible, or at least can't always be done in a clean way.

Reply:
I agree we generally want similar semantics where safe, but is it desirable to keep a similar syntax even when not idiomatic?  It might be, but I'm not sure if it is.

Reply:
We discussed this over and decided it's good when the checks are only debug assets so that arithmetic doesn't have hidden bounds checking costs.  The real bounds check still happens when actually indexing.